sklearn/cluster/_bisect_k_means.py

```diff
@@ -0,0 +1,687 @@
+
+
+    Attributes
+    ----------
+    cluster_centers_ : ndarray of shape (n_clusters, n_features)
+        Coordinates of cluster centers. If the algorithm stops before fully
+        converging (see ``tol`` and ``max_iter``), these will not be
+        consistent with ``labels_``.
+
+    labels_ : ndarray of shape (n_samples,)
+        Labels of each point
+
+    inertia_ : float
+        Sum of squared distances of samples to their closest cluster center,
+        weighted by the sample weights if provided.
+
+    n_iter_ : int
+        Number of iterations run.
+
+    Notes
+    -----
+    That algorithm will not work if n_cluster is smaller than 2.
+
+    Also it might be inefficient when n_cluster is equal to 2
+
+    Examples
+    --------
+    >>> from sklearn.cluster import BisectKMeans
+    >>> import numpy as np
+    >>> X = np.array([[1, 2], [1, 4], [1, 0],
+    ...               [10, 2], [10, 4], [10, 0],
+    ...               [10, 6], [10, 8], [10, 10]])
+    >>> bisect_means = BisectKMeans(n_clusters=3, random_state=0).fit(X)
+    >>> bisect_means.labels_
+    array([0, 0, 0, 2, 2, 2, 1, 1, 1], dtype=int32)
+    >>> bisect_means.predict([[0, 0], [12, 3]])
+    array([0, 2], dtype=int32)
+    >>> bisect_means.cluster_centers_
+    array([[ 1.,  2.],
+           [10.,  8.],
+           [10.,  2.]])
+    """
+    def __init__(self,  n_clusters=8, init='k-means++', n_init=10,
+                 random_state=None, max_iter=30, verbose=0,
+                 tol=1e-4, copy_x=True, algorithm='auto',
+                 bisect_strategy='biggest_sse'):
+
+        super().__init__(
+            n_clusters=n_clusters, init=init, max_iter=max_iter,
+            verbose=verbose, random_state=random_state, tol=tol,
+            n_init=n_init, copy_x=copy_x, algorithm=algorithm)
+
+        self.bisect_strategy = bisect_strategy
+
+    def _compute_bisect_errors(self, X, centers, labels):
+        """
+        Calculate the squared error of each sample and group them by label.
+
+        .. note:: That function works only if there are two labels (0,1) and
+        may be less efficient for sparse data.
+
+        Parameters
+        ----------
+        X : {ndarray, sparse matrix} of shape (n_samples, n_features)
+            The input samples.
+
+        centers : ndarray of shape (n_clusters, n_features)
+            The cluster centers.
+
+        labels : ndarray of shape (n_samples,)
+            Index of the cluster each sample belongs to.
+
+        Returns
+        -------
+        errors_by_label : dict
+            dictionary containing squared error of each point by label
+            as ndarray.
+        """
+        errors_by_label = {}
+
+        if sp.issparse(X):
+            for value in range(2):
+                # CSR Matrix after subtracting numpy array
+                # becomes numpy matrix
+                error = sp.csr_matrix(X[labels == value] - centers[value])
```

Comment:
This cast CSR -> numpy -> CSR is not great, since it might blow up the memory. I would use `_inertia_sparse` instead, which is doing this computation in the right way.

> Am I doing something wrong here? Because it returns me good values for error_by_label[0] but different at error_by_label[1] compared to _compute_bisect_errors 

You are giving an array `_labels` full of ones, where it should be an array full of zeros. You are giving only one cluster center at a time, so the function `_inertia_sparse` assumes there is only one cluster, so it only considers samples that are labeled with label = 0. 

It becomes:
```py

        if sp.issparse(X):
            for value in range(2):
                indexes = (labels == value)
                test = X[indexes]
                weights = sample_weight[indexes]

                _centers = centers[value][np.new_axis, :]
                _labels = np.zeros(test.shape[0])
                errors_by_label[value] = _inertia_sparse(test, weights,
                                                         _centers, _labels,
                                                         self._n_threads)
```

I would also use `_inertia_dense` which might be more faster or more numerically precise.

Reply:
That was more cause I had problems with` _inertia_sparse`. I'm not also into that solution, but now I may rewrite that with `_inertia_sparse`. Thank you for that